x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization/CategorizeTextAggregationBuilder.java

```diff
@@ -0,0 +1,275 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.aggs.categorization;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.ParseField;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.search.aggregations.AbstractAggregationBuilder;
+import org.elasticsearch.search.aggregations.AggregationBuilder;
+import org.elasticsearch.search.aggregations.AggregatorFactories;
+import org.elasticsearch.search.aggregations.AggregatorFactory;
+import org.elasticsearch.search.aggregations.bucket.terms.TermsAggregationBuilder;
+import org.elasticsearch.search.aggregations.bucket.terms.TermsAggregator;
+import org.elasticsearch.search.aggregations.support.AggregationContext;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class CategorizeTextAggregationBuilder extends AbstractAggregationBuilder<CategorizeTextAggregationBuilder> {
+
+    static final TermsAggregator.BucketCountThresholds DEFAULT_BUCKET_COUNT_THRESHOLDS = new TermsAggregator.BucketCountThresholds(
+        1,
+        0,
+        10,
+        -1
+    );
+    public static final String NAME = "categorize_text";
+
+    static final ParseField FIELD_NAME = new ParseField("field");
+    static final ParseField MAX_CHILDREN = new ParseField("max_children");
+    static final ParseField SIMILARITY_THRESHOLD = new ParseField("similarity_threshold");
+    static final ParseField MAX_DEPTH = new ParseField("max_depth");
+    static final ParseField CATEGORIZATION_FILTERS = new ParseField("categorization_filters");
+
+    public static final ObjectParser<CategorizeTextAggregationBuilder, String> PARSER = ObjectParser.fromBuilder(
+        CategorizeTextAggregationBuilder.NAME,
+        CategorizeTextAggregationBuilder::new
+    );
+    static {
+        PARSER.declareString(CategorizeTextAggregationBuilder::setFieldName, FIELD_NAME);
+        PARSER.declareInt(CategorizeTextAggregationBuilder::setMaxChildren, MAX_CHILDREN);
+        PARSER.declareInt(CategorizeTextAggregationBuilder::setMaxDepth, MAX_DEPTH);
```

Comment:
How deep can we go? I could see us allocating a ton of memory making a big, bushy tree here. I see that you are doing the ramBytesUsed thing so you can do some kind of accounting so it's not going to blow up, but maybe there should be a limit

Reply:
@nik9000 I agree, we should have a limit, I am just not 100% sure what it should be. I will think about it and do some investigation.